drivers/iommu/riscv/iommu.c

```diff
@@ -831,612 +1096,692 @@ static struct iommu_group *riscv_iommu_device_group(struct device *dev)
+	else
+		dc->tc = cpu_to_le64(RISCV_IOMMU_DC_TC_V);
+	ep->dc = dc;
+	mutex_unlock(&domain->lock);
+	riscv_iommu_iodir_inv_devid(ep->iommu, ep->devid);
 
-	return val;
+	// enable PCI capabilities
+	riscv_iommu_enable_pci_ep(dev);
+
+	return 0;
 }
 
-static int riscv_iommu_queue_init(struct riscv_iommu_device *iommu,
-				  struct riscv_iommu_queue *q, unsigned count,
-				  size_t item_size, unsigned id, unsigned qbr,
-				  unsigned qcr, irq_handler_t irq_fn,
-				  const char *name)
+static int riscv_iommu_set_dev_pasid(struct iommu_domain *iommu_domain,
+				     struct device *dev, ioasid_t pasid)
 {
-	struct device *dev = iommu->dev;
-	unsigned order = ilog2(count);
-
-	do {
-		size_t size = item_size * (1ULL << order);
-		q->base = dmam_alloc_coherent(dev, size, &q->base_dma,
-					      GFP_KERNEL);
-		if (q->base || size < PAGE_SIZE)
-			break;
-
-		order--;
-	} while (1);
+	struct riscv_iommu_domain *domain = iommu_domain_to_riscv(iommu_domain);
+	struct riscv_iommu_endpoint *ep = dev_iommu_priv_get(dev);
+	struct riscv_iommu_dc *dc = ep->dc;
+	struct riscv_iommu_pc *pc = ep->pc;
+	struct mm_struct *mm;
 
-	if (!q->base) {
-		dev_err(dev, "failed to allocate %s queue (cnt: %u)\n",
-			name, count);
+	if (!iommu_domain || !iommu_domain->mm)
+		return -EINVAL;
+	if (!ep || !ep->sva_supported)
+		return -ENODEV;
+	if (!pc)
+		pc = (struct riscv_iommu_pc *)get_zeroed_page(GFP_KERNEL);
+	if (!pc)
 		return -ENOMEM;
-	}
 
-	q->len = item_size;
-	q->cnt = 1ULL << order;
-	q->qbr = qbr;
-	q->qcr = qcr;
-
-	switch (FIELD_GET(RIO_CAP_IGS, iommu->cap)) {
-	case RIO_IGS_MSI:
-	case RIO_IGS_ANY:
-		q->irq = msi_get_virq(dev, id);
-		if (!q->irq) {
-			dev_warn(dev, "no MSI vector %d for %s\n", id, name);
-		} else if (request_threaded_irq(q->irq, NULL, irq_fn, IRQF_ONESHOT, dev_name(dev), q)) {
-			dev_warn(dev, "fail to request irq %d for %s\n", q->irq, name);
-			q->irq = 0;
-		}
-		break;
+	/* register mm notifier */
+	mm = iommu_domain->mm;
+	domain->pscid = pasid;
+	domain->iommu = ep->iommu;
+	domain->mn.ops = &riscv_iommu_mmuops;
 
-	case RIO_IGS_WSI:
-		dev_err(dev, "MSI support is missing.\n");
-		break;
+	if (mmu_notifier_register(&domain->mn, mm))
+	    return -ENODEV;
 
-	default:
-		dev_err(dev, "invalid interrupt generation support.\n");
-		break;
-	}
+	/* Use PASID for PSCID tag */
+	pc[pasid].ta = cpu_to_le64(FIELD_PREP(RISCV_IOMMU_PC_TA_PSCID, pasid) |
+		       RISCV_IOMMU_PC_TA_V);
+	/* TODO: get SXL value for the process, use 32 bit or SATP mode */
+	pc[pasid].fsc = cpu_to_le64(virt_to_pfn(mm->pgd) | satp_mode);
+	/* update DC with sva->mm */
+	if (!(ep->dc->tc & RISCV_IOMMU_DC_TC_PDTV)) {
+		/* migrate to PD, domain mappings moved to PASID:0 */
+		pc[0].ta = dc->ta;
+		pc[0].fsc = dc->fsc;
+		dc->fsc = cpu_to_le64(virt_to_pfn(pc) |
+			  FIELD_PREP(RISCV_IOMMU_DC_FSC_MODE, RISCV_IOMMU_DC_FSC_PDTP_MODE_PD8));
+		/* XXX: What is (1ULL << 32) ? it's in the custom region */
```

Comment:
This is temporary extension for QEMU IOMMU model to emulate PRI.
Will be removed before upstreaming (once QEMU complete ATS/PRI model is in place).

Reference: https://github.com/tjeznach/qemu/blob/tjeznach/riscv-iommu/hw/riscv/riscv-iommu.c#L1098